In [273]:
import pandas as pd
import pybaseball as pyb
import altair as alt

 ## **Introduction:** I am using @MLBPlayerAnalys Tommy John Surgery List Data Set (2438 lines) which contains information about MLB pitchers and Tommy John Surgery (UCL Surgery). I am curious about the characteristics associated with pitchers who undergo Tommy John surgery. I will use pybaseball data to get this pitch data to create a model to predict TJ surgery.
 
https://docs.google.com/spreadsheets/d/1gQujXQQGOVNaiuwSN680Hq-FDVsCwvN-3AazykOBON0/edit#gid=0

In [295]:
tj_data = pd.read_csv('/Users/zoeskigen/Desktop/tommyjohn.csv')

In [513]:
tj_data.head()

Player TJ Surgery Date Team Level Position Throws  \
0          Wade Miley      2024-05-01  MIL   MLB        P      L   
1   Robert Stephenson      2024-04-30  LAA   MLB        P      R   
2    Brandon Barriera      2024-04-29  TOR     A        P      L   
3  Sawyer Gipson-Long      2024-04-22  DET     A        P      R   
4        Shane Bieber      2024-04-12  CLE   MLB        P      R   

         Country High School                College(s)   Age  \
0  United States   Louisiana    Southeastern Louisiana  37.0   
1  United States  California                       NaN  31.0   
2  United States     Florida                       NaN  20.0   
3  United States     Georgia                    Mercer  26.0   
4  United States  California  California Santa Barbara  29.0   

  Return Date (same level)  Recovery Time (months) mlbamid       fgid  \
0                      NaN                     NaN  489119       8779   
1                      NaN                     NaN  596112      13594   
2                      NaN                     NaN  702255  sa3020683   
3                      NaN                     NaN  687830      26048   
4                      NaN                     NaN  669456      19427   

          Surgeon(s)  Year  Month  
0                NaN  2024      5  
1                NaN  2024      4  
2  Dr. Keith Meister  2024      4  
3                NaN  2024      4  
4  Dr. Keith Meister  2024      4

In [515]:
tj_data.columns

Index(['Player', 'TJ Surgery Date', 'Team', 'Level', 'Position', 'Throws',
       'Country', 'High School', 'College(s)', 'Age',
       'Return Date (same level)', 'Recovery Time (months)', 'mlbamid', 'fgid',
       'Surgeon(s)', 'Year', 'Month'],
      dtype='object')

In [517]:
columns = [
    'Player', 'TJ Surgery Date', 'Team', 'Level', 'Position', 'Throws', 
    'Country', 'High School', 'College(s)', 'Age', 'Return Date (same level)', 
    'Recovery Time (months)', 'mlbamid', 'fgid', 'Surgeon(s)'
]

# Select only the columns you want to keep from tj_data
tj_data = tj_data[columns]

In [ ]:
tj_data

In [521]:
# Ensure 'TJ Surgery Date' is converted to datetime and extract the year
tj_data['TJ Surgery Date'] = pd.to_datetime(tj_data['TJ Surgery Date'], errors='coerce')
tj_data['Year'] = tj_data['TJ Surgery Date'].dt.year

# Filter the DataFrame for MLB level, exclude 2024, and include years 2000 and later
mlb_data = tj_data[(tj_data['Level'] == 'MLB') & (tj_data['Year'] >= 2000) & (tj_data['Year'] < 2024)]

# Group by 'Year' and count the number of surgeries for MLB pitchers
mlb_surgeries_by_year = mlb_data.groupby('Year').size().reset_index(name='Count')

# Create a line graph of the total number of MLB surgeries by year, excluding 2024
mlb_surgeries_trend_chart = alt.Chart(mlb_surgeries_by_year).mark_line(point=True).encode(
    alt.X('Year:O', title='Year'),
    alt.Y('Count:Q', title='Total Number of Surgeries'),
    tooltip=[alt.Tooltip('Year:O', title='Year'), alt.Tooltip('Count:Q', title='Total Surgeries')]
).properties(
    title='Total Number of Tommy John Surgeries for MLB Pitchers from 2000 to 2023'
)

mlb_surgeries_trend_chart.display()

alt.Chart(...)

## Total Number of TJ Surgeries increasing over the years

In [523]:
# Create a histogram of ages for all pitchers
all_pitchers_age_chart = alt.Chart(tj_data).mark_bar().encode(
    alt.X('Age:Q', bin=True, title='Age'),
    alt.Y('count()', title='Number of Pitchers'),
    tooltip=[alt.Tooltip('count()', title='Number of Pitchers'), 'Age']
).properties(
    title='Age Distribution of All Pitchers Undergoing Tommy John Surgery'
)

all_pitchers_age_chart.display()

alt.Chart(...)

## Average Age to get TJ is around 24 years old

In [525]:
# Filter the DataFrame for years 2000 and later
tj_data = tj_data[tj_data['Year'] >= 2000]
# Group by 'Year' and calculate the average age
average_age_by_year = tj_data.groupby('Year')['Age'].mean().reset_index()

age_trend_chart = alt.Chart(average_age_by_year).mark_line(point=True).encode(
    alt.X('Year:O', title='Year'),
    alt.Y('Age:Q', title='Average Age'),
    tooltip=[alt.Tooltip('Year:O', title='Year'), alt.Tooltip('Age:Q', title='Average Age')]
).properties(
    title='Trend of Average Age of Pitchers Undergoing Tommy John Surgery from 2000 Onwards'
)

age_trend_chart.display()



alt.Chart(...)

## Average age of pitchers getting TJ is increasing, could be because average age of MLB debut is increasing?

In [527]:
# Ensure 'TJ Surgery Date' is converted to datetime and extract the month correctly
tj_data['TJ Surgery Date'] = pd.to_datetime(tj_data['TJ Surgery Date'], errors='coerce')
tj_data['Month'] = tj_data['TJ Surgery Date'].dt.month

# Re-filter the DataFrame for MLB level if needed
mlb_data = tj_data[tj_data['Level'] == 'MLB']

# Create a bar chart of the total number of surgeries by month
surgeries_month_chart = alt.Chart(surgeries_by_month).mark_bar().encode(
    alt.X('Month:N', title='Month', 
          scale=alt.Scale(domain=list(range(1, 13))),  # Ensure the domain covers all months
          axis=alt.Axis(values=list(range(1, 13)), labelExpr="['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'][datum.value - 1]")
         ),
    alt.Y('Count:Q', title='Total Number of Surgeries'),
    tooltip=[alt.Tooltip('Month:N', title='Month'), alt.Tooltip('Count:Q', title='Total Surgeries')]
).properties(
    title='Total Number of Tommy John Surgeries by Month for MLB Pitchers',
    width=600,
    height=400
)

# Display the chart
surgeries_month_chart.display()

alt.Chart(...)

## Not a massive difference in TJ surgeries occuring throughout the season. But the bulk occur within the first 4 months of the season.

# **Getting MLB pitch characteristics from PyBaseball, gives information from Statcast**

In [529]:
try:
    pitching_data_2023 = pyb.statcast('2023-03-01', '2023-10-01')
    print("2023 data fetched successfully.")
except Exception as e:
    print(f"Error fetching 2023 data: {e}")

try:
    pitching_data_2024 = pyb.statcast('2024-03-01', '2024-10-01')
    print("2024 data fetched successfully.")
except Exception as e:
    print(f"Error fetching 2024 data: {e}")

This is a large query, it may take a moment to complete
Skipping offseason dates


100%|██████████| 201/201 [00:32<00:00,  6.20it/s]
/Users/zoeskigen/anaconda3/lib/python3.11/site-packages/pybaseball/statcast.py:85: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  final_data = pd.concat(dataframe_list, axis=0).convert_dtypes(convert_string=False)


2023 data fetched successfully.
This is a large query, it may take a moment to complete
Skipping offseason dates


100%|██████████| 201/201 [00:33<00:00,  6.04it/s]
/Users/zoeskigen/anaconda3/lib/python3.11/site-packages/pybaseball/statcast.py:85: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  final_data = pd.concat(dataframe_list, axis=0).convert_dtypes(convert_string=False)


2024 data fetched successfully.


In [531]:
# Assuming you have both 2023 and 2024 data
if 'pitching_data_2023' in locals() and not pitching_data_2023.empty:
    pitching_data = pd.concat([pitching_data_2023, pitching_data_2024], ignore_index=True)
    print("Data from 2023 and 2024 combined successfully.")
else:
    pitching_data = pitching_data_2024
    print("Using only 2024 data.")

Data from 2023 and 2024 combined successfully.


In [541]:
pitching_data.head()

pitch_type  game_date  release_speed  release_pos_x  release_pos_z  \
0         CH 2023-10-01           89.0           -2.8           5.59   
1         FF 2023-10-01           96.9           -2.4            5.9   
2         CH 2023-10-01           90.0          -2.93           5.56   
3         ST 2023-10-01           82.2          -3.09           5.55   
4         CH 2023-10-01           89.2          -2.87           5.58   

       player_name  batter mlbamid     events      description  ...  \
0  Robertson, Nick  677008  687798  field_out    hit_into_play  ...   
1  Robertson, Nick  677008  687798       None             foul  ...   
2  Robertson, Nick  677008  687798       None             ball  ...   
3  Robertson, Nick  677008  687798       None             ball  ...   
4  Robertson, Nick  677008  687798       None  swinging_strike  ...   

   fld_score  post_away_score  post_home_score  post_bat_score  \
0          6                6                1               1   
1          6                6                1               1   
2          6                6                1               1   
3          6                6                1               1   
4          6                6                1               1   

   post_fld_score if_fielding_alignment of_fielding_alignment spin_axis  \
0               6         Infield shade              Standard       250   
1               6         Infield shade              Standard       211   
2               6         Infield shade              Standard       250   
3               6         Infield shade              Standard        58   
4               6         Infield shade              Standard       257   

  delta_home_win_exp delta_run_exp  
0                0.0         -0.09  
1                0.0           0.0  
2                0.0         0.016  
3                0.0         0.009  
4                0.0        -0.023  

[5 rows x 92 columns]

In [543]:
pitching_data.rename(columns={'pitcher': 'mlbamid'}, inplace=True)


In [545]:
tj_data['mlbamid'] = tj_data['mlbamid'].astype(str)
pitching_data['mlbamid'] = pitching_data['mlbamid'].astype(str)

In [549]:
# Merge the data using 'mlbamid'
combined_data = pd.merge(tj_data, pitching_data, on='mlbamid', how='inner')
print("Data merged successfully.")
print(combined_data.head())

Data merged successfully.
       Player TJ Surgery Date Team Level Position Throws        Country  \
0  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
1  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
2  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
3  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
4  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   

  High School              College(s)   Age  ... fld_score  post_away_score  \
0   Louisiana  Southeastern Louisiana  37.0  ...         0                1   
1   Louisiana  Southeastern Louisiana  37.0  ...         0                1   
2   Louisiana  Southeastern Louisiana  37.0  ...         0                1   
3   Louisiana  Southeastern Louisiana  37.0  ...         0                1   
4   Louisiana  Southeastern Louisiana  37.0  ...         0                1   

  post_home_score post_bat_score post_fld_score 

In [551]:
common_ids = tj_data[tj_data['mlbamid'].isin(pitching_data['mlbamid'])]
print("Number of common MLBAM IDs:", len(common_ids))
print(common_ids['mlbamid'].head())


Number of common MLBAM IDs: 476
0    489119
1    596112
3    687830
4    669456
5    669684
Name: mlbamid, dtype: object


In [553]:
combined_data.columns

Index(['Player', 'TJ Surgery Date', 'Team', 'Level', 'Position', 'Throws',
       'Country', 'High School', 'College(s)', 'Age',
       ...
       'fld_score', 'post_away_score', 'post_home_score', 'post_bat_score',
       'post_fld_score', 'if_fielding_alignment', 'of_fielding_alignment',
       'spin_axis', 'delta_home_win_exp', 'delta_run_exp'],
      dtype='object', length=108)

## Get the data we actually want - pitch characteristics like spin rate, release point, etc

In [555]:
# List of relevant pitch characteristics columns
pitch_columns = [
    'pitch_type', 'release_speed', 'release_spin_rate', 
    'release_pos_x', 'release_pos_z', 'spin_axis', 'effective_speed', 'release_extension'
] #can add more later if i want

# Add any other columns from tj_data that are relevant to your analysis
relevant_tj_columns = [
    'Player', 'TJ Surgery Date', 'Team', 'Level', 'Position', 'Throws', 
    'Country', 'High School', 'College(s)', 'Age', 'mlbamid'
]

# Combine the lists if needed
all_relevant_columns = relevant_tj_columns + pitch_columns

In [557]:
# Filter the DataFrame to include only relevant columns
filtered_combined_data = combined_data[all_relevant_columns]
filtered_combined_data

Player TJ Surgery Date Team Level Position Throws  \
0         Wade Miley      2024-05-01  MIL   MLB        P      L   
1         Wade Miley      2024-05-01  MIL   MLB        P      L   
2         Wade Miley      2024-05-01  MIL   MLB        P      L   
3         Wade Miley      2024-05-01  MIL   MLB        P      L   
4         Wade Miley      2024-05-01  MIL   MLB        P      L   
...              ...             ...  ...   ...      ...    ...   
375354  Wily Peralta      2007-01-01  MIL    Rk        P      R   
375355  Wily Peralta      2007-01-01  MIL    Rk        P      R   
375356  Wily Peralta      2007-01-01  MIL    Rk        P      R   
375357  Wily Peralta      2007-01-01  MIL    Rk        P      R   
375358  Wily Peralta      2007-01-01  MIL    Rk        P      R   

              Country High School              College(s)   Age mlbamid  \
0       United States   Louisiana  Southeastern Louisiana  37.0  489119   
1       United States   Louisiana  Southeastern Louisiana  37.0  489119   
2       United States   Louisiana  Southeastern Louisiana  37.0  489119   
3       United States   Louisiana  Southeastern Louisiana  37.0  489119   
4       United States   Louisiana  Southeastern Louisiana  37.0  489119   
...               ...         ...                     ...   ...     ...   
375354      Dominican         NaN                     NaN  18.0  503449   
375355      Dominican         NaN                     NaN  18.0  503449   
375356      Dominican         NaN                     NaN  18.0  503449   
375357      Dominican         NaN                     NaN  18.0  503449   
375358      Dominican         NaN                     NaN  18.0  503449   

       pitch_type  release_speed  release_spin_rate  release_pos_x  \
0              FC           87.9               2296           1.67   
1              FC           87.3               2143           1.87   
2              FC           87.2               2065           1.73   
3              CH           84.1               1749           2.13   
4              FC           87.9               2066           1.83   
...           ...            ...                ...            ...   
375354         FS           77.9                947          -1.32   
375355         FS           78.0               1291          -1.44   
375356         FF           91.7               1932          -1.63   
375357         FS           78.0               1348           -1.4   
375358         FF           92.5               1974          -1.42   

        release_pos_z  spin_axis  effective_speed  release_extension  
0                6.27        191             87.8                6.1  
1                6.21        174             87.2                6.2  
2                6.25        173             87.1                6.1  
3                5.99        131             83.8                6.2  
4                6.27        176             88.2                6.2  
...               ...        ...              ...                ...  
375354           6.46        226             77.2                6.0  
375355           6.41        220             77.7                6.0  
375356           6.36        214             90.8                5.7  
375357            6.5        218             77.5                6.0  
375358           6.52        206             91.5                5.8  

[375359 rows x 19 columns]

In [559]:
print(filtered_combined_data.columns)

Index(['Player', 'TJ Surgery Date', 'Team', 'Level', 'Position', 'Throws',
       'Country', 'High School', 'College(s)', 'Age', 'mlbamid', 'pitch_type',
       'release_speed', 'release_spin_rate', 'release_pos_x', 'release_pos_z',
       'spin_axis', 'effective_speed', 'release_extension'],
      dtype='object')


In [561]:
# Use .loc to ensure the operation modifies the DataFrame directly
filtered_combined_data.loc[:, 'game_year'] = pd.to_datetime(filtered_combined_data['TJ Surgery Date']).dt.year

/var/folders/mj/b8_lt8gn6cvcflsf00z0pn6r0000gn/T/ipykernel_4367/800781858.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_combined_data.loc[:, 'game_year'] = pd.to_datetime(filtered_combined_data['TJ Surgery Date']).dt.year


In [563]:
filtered_combined_data.head()

Player TJ Surgery Date Team Level Position Throws        Country  \
0  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
1  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
2  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
3  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   
4  Wade Miley      2024-05-01  MIL   MLB        P      L  United States   

  High School              College(s)   Age mlbamid pitch_type  release_speed  \
0   Louisiana  Southeastern Louisiana  37.0  489119         FC           87.9   
1   Louisiana  Southeastern Louisiana  37.0  489119         FC           87.3   
2   Louisiana  Southeastern Louisiana  37.0  489119         FC           87.2   
3   Louisiana  Southeastern Louisiana  37.0  489119         CH           84.1   
4   Louisiana  Southeastern Louisiana  37.0  489119         FC           87.9   

   release_spin_rate  release_pos_x  release_pos_z  spin_axis  \
0               2296           1.67           6.27        191   
1               2143           1.87           6.21        174   
2               2065           1.73           6.25        173   
3               1749           2.13           5.99        131   
4               2066           1.83           6.27        176   

   effective_speed  release_extension  game_year  
0             87.8                6.1       2024  
1             87.2                6.2       2024  
2             87.1                6.1       2024  
3             83.8                6.2       2024  
4             88.2                6.2       2024

## Filter for MLB level players' fastballs

In [565]:
# Filter for MLB level and pitchers who underwent Tommy John surgery
mlb_surgery_pitchers = filtered_combined_data[
    (filtered_combined_data['Level'] == 'MLB') & 
    (filtered_combined_data['TJ Surgery Date'].notna())
]

# Further filter for fastballs, assuming 'FF' denotes four-seam fastballs
mlb_fastballs_after_surgery = mlb_surgery_pitchers[mlb_surgery_pitchers['pitch_type'] == 'FF']

In [567]:
# Group by year and calculate the average velocity
average_velocity_by_year = mlb_fastballs_after_surgery.groupby('game_year')['release_speed'].mean()
print("Average Fastball Velocity by Year for MLB Pitchers Who Underwent TJ Surgery:")
print(average_velocity_by_year)

Average Fastball Velocity by Year for MLB Pitchers Who Underwent TJ Surgery:
game_year
2011    89.538957
2012    94.689179
2013    95.271069
2014    93.840439
2015    94.054738
2016    93.209905
2017    93.394571
2018    94.231029
2019    94.648529
2020    94.541888
2021    94.441203
2022    94.002903
2023    95.342525
2024    94.064952
Name: release_speed, dtype: Float64


In [569]:
# Define the scale range based on the data
min_velocity = mlb_fastballs_after_surgery['release_speed'].min()
max_velocity = mlb_fastballs_after_surgery['release_speed'].max()

# Create a line chart to visualize average fastball velocity over time
velocity_chart = alt.Chart(average_velocity_by_year.reset_index()).mark_line(point=True).encode(
    x='game_year:O',
    y=alt.Y('release_speed:Q', scale=alt.Scale(domain=[min_velocity, max_velocity])),
    tooltip=['game_year', 'release_speed']
).properties(
    title='Average Fastball Velocity Over Time for MLB Pitchers Who Underwent Tommy John Surgery',
    width=600,
    height=300
)

velocity_chart.display()

alt.Chart(...)

## Not as significant change as I would have expected considering pitchers are capable of pitching 100+ mph, maybe because those with the highest velo are relief pitchers and the less significant workload makes them less suspectible to TJ.

In [571]:
# Convert Series to DataFrame for Altair
data = pitch_counts.reset_index()
data.columns = ['pitch_type', 'count']

# Create a bar chart
chart = alt.Chart(data).mark_bar().encode(
    x='pitch_type',
    y='count',
    tooltip=['pitch_type', 'count']
).properties(
    title='Distribution of Pitch Types among MLB Players with TJ Surgery'
)

chart.display()

alt.Chart(...)

## Most common pitches amongst TJ pitchers: 4-Seam Fastball (by far), Slider, Sinker

In [573]:
# Assuming 'filtered_combined_data' is your DataFrame
mlb_data = filtered_combined_data[
    (filtered_combined_data['Level'] == 'MLB') & 
    (filtered_combined_data['TJ Surgery Date'].notna())
]

# Group by 'pitch_type' and calculate count and average 'release_speed'
pitch_stats = mlb_data.groupby('pitch_type').agg({
    'release_speed': 'mean',  # Average velocity
    'pitch_type': 'size'      # Count of each pitch type
}).rename(columns={'release_speed': 'average_velocity', 'pitch_type': 'count'}).reset_index()

# Display the DataFrame to check it
print(pitch_stats)

   pitch_type  average_velocity  count
0          CH          85.67096  15653
1          CS         66.843478     23
2          CU         78.384989  12129
3          FC         88.330487  14442
4          FF          94.20226  52940
5          FS         86.551397   7481
6          KC         81.047687   3416
7          PO             90.78     10
8          SI         93.239209  26706
9          SL         84.889718  27626
10         ST         81.536225   5568
11         SV              81.6      1


In [575]:
scatter_chart = alt.Chart(pitch_stats).mark_circle(size=100).encode(
    x=alt.X('count:Q', title='Distribution of Pitch Types'),
    y=alt.Y('average_velocity:Q', title='Average Velocity (mph)'),
    tooltip=['pitch_type:N', 'count:Q', 'average_velocity:Q'],
    color='pitch_type:N'
).properties(
    title='Average Velocity vs Distribution of Pitch Types among MLB Players with TJ Surgery'
)

scatter_chart.display()

alt.Chart(...)

## ** Find data for MLB pitchers who have NOT undergone TJ surgery ** 

In [577]:
# Fetch MLB pitching data for 2023
mlb_pitching_data_2023 = pyb.statcast('2023-04-01', '2023-10-01')

# If you are fetching this in real-time and 2024 data is available
try:
    mlb_pitching_data_2024 = pyb.statcast('2024-04-01', '2024-10-01')
except Exception as e:
    print("2024 data not available yet:", e)
    mlb_pitching_data_2024 = None

This is a large query, it may take a moment to complete


100%|██████████| 184/184 [00:29<00:00,  6.17it/s]
/Users/zoeskigen/anaconda3/lib/python3.11/site-packages/pybaseball/statcast.py:85: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  final_data = pd.concat(dataframe_list, axis=0).convert_dtypes(convert_string=False)


This is a large query, it may take a moment to complete


100%|██████████| 184/184 [00:31<00:00,  5.80it/s]
/Users/zoeskigen/anaconda3/lib/python3.11/site-packages/pybaseball/statcast.py:85: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  final_data = pd.concat(dataframe_list, axis=0).convert_dtypes(convert_string=False)


In [579]:
if mlb_pitching_data_2024 is not None:
    mlb_pitching_data = pd.concat([mlb_pitching_data_2023, mlb_pitching_data_2024], ignore_index=True)
else:
    mlb_pitching_data = mlb_pitching_data_2023

In [581]:
tj_pitcher_ids = set(tj_data['mlbamid'].dropna().astype(str))

In [583]:
# Filter out pitchers who have had TJ surgery using the 'pitcher' column
non_tj_mlb_pitching_data = mlb_pitching_data[~mlb_pitching_data['pitcher'].astype(str).isin(tj_pitcher_ids)]


In [587]:
# Filter out pitchers who have had TJ surgery using the 'pitcher' column
non_tj_mlb_pitching_data = mlb_pitching_data[~mlb_pitching_data['pitcher'].astype(str).isin(tj_pitcher_ids)]

In [589]:
print("Number of entries in original MLB data:", len(mlb_pitching_data))
print("Number of entries after excluding TJ surgery pitchers:", len(non_tj_mlb_pitching_data))

Number of entries in original MLB data: 832542
Number of entries after excluding TJ surgery pitchers: 518937


In [591]:
# Filter for fastballs, assuming 'FF' denotes four-seam fastballs
non_tj_fastballs = non_tj_mlb_pitching_data[non_tj_mlb_pitching_data['pitch_type'] == 'FF']

# Group by year and calculate the average velocity
average_velocity_by_year = non_tj_fastballs.groupby('game_year')['release_speed'].mean()
print("Average Fastball Velocity by Year for Non-TJ MLB Pitchers:")
print(average_velocity_by_year)

Average Fastball Velocity by Year for Non-TJ MLB Pitchers:
game_year
2023    94.082255
2024    93.769582
Name: release_speed, dtype: Float64


In [593]:
non_tj_mlb_pitching_data 

pitch_type  game_date  release_speed  release_pos_x  release_pos_z  \
29             ST 2023-10-01           80.3          -2.94           3.72   
30             ST 2023-10-01           80.1           -2.9           3.95   
31             FF 2023-10-01           93.7          -2.66           4.31   
32             FF 2023-10-01           94.0          -2.77            4.2   
33             FF 2023-10-01           94.6          -2.57           4.24   
...           ...        ...            ...            ...            ...   
832537         FF 2024-04-01           97.5           1.77           5.84   
832538         FF 2024-04-01           98.6           2.02           5.88   
832539         FF 2024-04-01           97.7           2.05           5.83   
832540         FF 2024-04-01           96.8            2.1           5.84   
832541         FF 2024-04-01           97.3           2.14           5.78   

            player_name  batter  pitcher     events    description  ...  \
29      Schreiber, John  623993   670167  field_out  hit_into_play  ...   
30      Schreiber, John  623993   670167       None           ball  ...   
31      Schreiber, John  623993   670167       None           ball  ...   
32      Schreiber, John  668939   670167     single  hit_into_play  ...   
33      Schreiber, John  668939   670167       None           foul  ...   
...                 ...     ...      ...        ...            ...  ...   
832537  Gore, MacKenzie  656582   669022       None           ball  ...   
832538  Gore, MacKenzie  656582   669022       None           ball  ...   
832539  Gore, MacKenzie  656582   669022       None           foul  ...   
832540  Gore, MacKenzie  656582   669022       None           ball  ...   
832541  Gore, MacKenzie  656582   669022       None  called_strike  ...   

        fld_score  post_away_score  post_home_score  post_bat_score  \
29              6                6                1               1   
30              6                6                1               1   
31              6                6                1               1   
32              6                6                1               1   
33              6                6                0               0   
...           ...              ...              ...             ...   
832537          0                0                0               0   
832538          0                0                0               0   
832539          0                0                0               0   
832540          0                0                0               0   
832541          0                0                0               0   

        post_fld_score if_fielding_alignment of_fielding_alignment spin_axis  \
29                   6         Infield shade              Standard        67   
30                   6         Infield shade              Standard        67   
31                   6         Infield shade              Standard       240   
32                   6              Standard              Standard       237   
33                   6              Standard              Standard       234   
...                ...                   ...                   ...       ...   
832537               0              Standard              Standard       153   
832538               0              Standard              Standard       147   
832539               0              Standard              Standard       150   
832540               0              Standard              Standard       149   
832541               0              Standard              Standard       146   

       delta_home_win_exp delta_run_exp  
29                 -0.005        -0.297  
30                    0.0         0.031  
31                    0.0         0.025  
32                  0.005         0.848  
33                    0.0        -0.071  
...                   ...           ...  
832537                0.0         0.103  
832538                0.0      

In [595]:
# Filter for MLB pitchers and fastballs ('FF' for four-seam fastball) in filtered_combined_data
tj_fastballs = filtered_combined_data[
    (filtered_combined_data['TJ Surgery Date'].notna()) &
    (filtered_combined_data['pitch_type'] == 'FF')
]

# Assuming all entries in non_tj_mlb_pitching_data are MLB players
non_tj_fastballs = non_tj_mlb_pitching_data[
    (non_tj_mlb_pitching_data['pitch_type'] == 'FF')
]

# Calculate average fastball velocity
avg_velocity_tj = tj_fastballs['release_speed'].mean()
avg_velocity_non_tj = non_tj_fastballs['release_speed'].mean()

# Create a DataFrame for plotting
velocity_comparison = pd.DataFrame({
    'Group': ['TJ Surgery', 'No TJ Surgery'],
    'Average Velocity': [avg_velocity_tj, avg_velocity_non_tj]
})

# Create a bar chart for average fastball velocities
velocity_chart = alt.Chart(velocity_comparison).mark_bar().encode(
    x='Group',
    y='Average Velocity',
    color='Group',
    tooltip=['Group', 'Average Velocity']
).properties(
    title='Comparison of Average Fastball Velocity: TJ Surgery vs No TJ Surgery'
)

velocity_chart.display()

alt.Chart(...)

## The difference in velocity for pitchers undergoing TJ vs not is slight...

In [597]:
# Filter for MLB pitchers in both datasets
tj_mlb_pitchers = filtered_combined_data[
    (filtered_combined_data['Level'] == 'MLB') &
    (filtered_combined_data['TJ Surgery Date'].notna())
]

non_tj_mlb_pitchers = non_tj_mlb_pitching_data[
    (non_tj_mlb_pitching_data['pitch_type'].notna())  # Assuming all are MLB level if no 'Level' column
]

# Calculate average release extension
avg_extension_tj = tj_mlb_pitchers['release_extension'].mean()
avg_extension_non_tj = non_tj_mlb_pitchers['release_extension'].mean()

# Create a DataFrame for plotting
extension_comparison = pd.DataFrame({
    'Group': ['TJ Surgery', 'No TJ Surgery'],
    'Average Extension': [avg_extension_tj, avg_extension_non_tj]
})


In [599]:
extension_chart = alt.Chart(extension_comparison).mark_bar().encode(
    x='Group',
    y='Average Extension',
    color='Group',
    tooltip=['Group', 'Average Extension']
).properties(
    title='Comparison of Average Release Extension: TJ Surgery vs No TJ Surgery'
)

extension_chart.display()

alt.Chart(...)

## There is a little more of a noticible difference here. TJ surgery pitchers average 6.47 inches of extension while non TJ pitchers average 6.42 inches. While this difference may seem minor... baseball is a game of inches. Even a half an inch can change the trajectory of the game or a player.

In [651]:
from pybaseball import statcast

# Fetching statcast pitching data for the year 2023
# Note: Adjust the date range as needed based on data availability and your specific needs
pitching_data_2023 = statcast(start_dt='2023-01-01', end_dt='2023-12-31')

# For 2024, ensure the data is available and adjust dates accordingly
# pitching_data_2024 = statcast(start_dt='2024-01-01', end_dt='2024-12-31')

# If you fetched data for both years and both are available, concatenate them
# pitching_data = pd.concat([pitching_data_2023, pitching_data_2024], ignore_index=True)

This is a large query, it may take a moment to complete
Skipping offseason dates
Skipping offseason dates


100%|██████████| 246/246 [00:43<00:00,  5.70it/s]
/Users/zoeskigen/anaconda3/lib/python3.11/site-packages/pybaseball/statcast.py:85: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  final_data = pd.concat(dataframe_list, axis=0).convert_dtypes(convert_string=False)


In [669]:
# Create 'had_surgery' column based on 'TJ Surgery Date'
filtered_combined_data['had_surgery'] = filtered_combined_data['TJ Surgery Date'].notna()

/var/folders/mj/b8_lt8gn6cvcflsf00z0pn6r0000gn/T/ipykernel_4367/239044774.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_combined_data['had_surgery'] = filtered_combined_data['TJ Surgery Date'].notna()


In [671]:
print(filtered_combined_data[['Player', 'TJ Surgery Date', 'had_surgery']].head())

                  Player TJ Surgery Date  had_surgery
0             Wade Miley      2024-05-01         True
2073   Robert Stephenson      2024-04-30         True
2878  Sawyer Gipson-Long      2024-04-22         True
3214        Shane Bieber      2024-04-12         True
5534        Chris Murphy      2024-04-10         True


In [677]:
all_pitchers_data.columns

Index(['pitch_type', 'game_date', 'release_speed', 'release_pos_x',
       'release_pos_z', 'player_name', 'batter', 'pitcher', 'events',
       'description', 'spin_dir', 'spin_rate_deprecated',
       'break_angle_deprecated', 'break_length_deprecated', 'zone', 'des',
       'game_type', 'stand', 'p_throws', 'home_team', 'away_team', 'type',
       'hit_location', 'bb_type', 'balls', 'strikes', 'game_year', 'pfx_x',
       'pfx_z', 'plate_x', 'plate_z', 'on_3b', 'on_2b', 'on_1b',
       'outs_when_up', 'inning', 'inning_topbot', 'hc_x', 'hc_y',
       'tfs_deprecated', 'tfs_zulu_deprecated', 'fielder_2', 'umpire', 'sv_id',
       'vx0', 'vy0', 'vz0', 'ax', 'ay', 'az', 'sz_top', 'sz_bot',
       'hit_distance_sc', 'launch_speed', 'launch_angle', 'effective_speed',
       'release_spin_rate', 'release_extension', 'game_pk', 'pitcher.1',
       'fielder_2.1', 'fielder_3', 'fielder_4', 'fielder_5', 'fielder_6',
       'fielder_7', 'fielder_8', 'fielder_9', 'release_pos_y',
       'estima

In [675]:
# Ensure data types are appropriate for merging
pitching_data_2023['pitcher'] = pitching_data_2023['pitcher'].astype(str)
filtered_combined_data['mlbamid'] = filtered_combined_data['mlbamid'].astype(str)

# Merge the datasets
all_pitchers_data = pd.merge(
    pitching_data_2023,
    filtered_combined_data[['mlbamid', 'TJ Surgery Date', 'had_surgery']],
    left_on='pitcher',
    right_on='mlbamid',
    how='left'
)

# Fill NaN for players without known TJ surgery
all_pitchers_data['TJ Surgery Date'].fillna('No Surgery', inplace=True)
all_pitchers_data['had_surgery'].fillna(False, inplace=True)

all_pitchers_data

/var/folders/mj/b8_lt8gn6cvcflsf00z0pn6r0000gn/T/ipykernel_4367/4191295702.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_combined_data['mlbamid'] = filtered_combined_data['mlbamid'].astype(str)
/var/folders/mj/b8_lt8gn6cvcflsf00z0pn6r0000gn/T/ipykernel_4367/4191295702.py:15: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'No Surgery' has dtype incompatible with datetime64[ns], please explicitly cast to a compatible dtype first.
  all_pitchers_data['TJ Surgery Date'].fillna('No Surgery', inplace=True)


Merged Data Sample:


pitch_type  game_date  release_speed  release_pos_x  release_pos_z  \
0              CU 2023-11-01           84.9          -1.19           6.12   
1              FF 2023-11-01           96.6          -0.69           6.24   
2              CU 2023-11-01           84.5          -1.27           6.11   
3              FF 2023-11-01           95.4           -0.8           6.23   
4              FF 2023-11-01           95.6          -0.83           6.19   
...           ...        ...            ...            ...            ...   
771052       None 2023-03-15           <NA>           <NA>           <NA>   
771053       None 2023-03-15           <NA>           <NA>           <NA>   
771054       None 2023-03-15           <NA>           <NA>           <NA>   
771055       None 2023-03-15           <NA>           <NA>           <NA>   
771056       None 2023-03-15           <NA>           <NA>           <NA>   

          player_name  batter pitcher     events      description  ...  \
0         Sborz, Josh  606466  622250  strikeout    called_strike  ...   
1         Sborz, Josh  606466  622250       None             ball  ...   
2         Sborz, Josh  606466  622250       None    called_strike  ...   
3         Sborz, Josh  606466  622250       None             foul  ...   
4         Sborz, Josh  606466  622250       None             ball  ...   
...               ...     ...     ...        ...              ...  ...   
771052  Dunning, Dane  679845  641540  field_out    hit_into_play  ...   
771053  Dunning, Dane  668472  641540  strikeout  swinging_strike  ...   
771054  Dunning, Dane  668472  641540       None    called_strike  ...   
771055  Dunning, Dane  668472  641540       None             foul  ...   
771056  Dunning, Dane  681987  641540     double    hit_into_play  ...   

        post_bat_score  post_fld_score  if_fielding_alignment  \
0                    0               5          Infield shade   
1                    0               5              Strategic   
2                    0               5               Standard   
3                    0               5               Standard   
4                    0               5               Standard   
...                ...             ...                    ...   
771052               0               0                   None   
771053               0               0                   None   
771054               0               0                   None   
771055               0               0                   None   
771056               0               0                   None   

        of_fielding_alignment  spin_axis delta_home_win_exp delta_run_exp  \
0                    Standard         26                0.0         -0.09   
1                    Standard        206                0.0         0.016   
2                    Standard         24                0.0        -0.027   
3                    Standard        204                0.0        -0.022   
4                    Standard        207                0.0         0.016   
...                       ...        ...                ...           ...   
771052                   None       <NA>               0.03          <NA>   
771053                   None       <NA>              0.041          <NA>   
771054                   None       <NA>                0.0          <NA>   
771055                   None       <NA>                0.0          <NA>   
771056                   None       <NA>              -0.06          <NA>   

       mlbamid      TJ Surgery Date had_surgery  
0          NaN           No Surgery       False  
1          NaN           No Surgery       False  
2          NaN           No Surgery       False  
3          NaN           No Surgery       False  
4          NaN           No Surgery       False  
...        ...                  ...         ...  
771052  641540  2019-03-18 00:00:00        True  
771053  641540  2019-03-18 00:00:00        True  
771054  641540  2019-03-18 00:00:00  

In [876]:
pitchers_data = all_pitchers_data[['pitch_type', 'release_speed', 'release_pos_x', 'release_pos_z', 
                                  'pitcher', 'p_throws', 'release_spin_rate', 
                                   'effective_speed', 'inning', 'pitch_number', 'mlbamid', 
                                   'TJ Surgery Date', 'had_surgery', 'game_date', 'game_year']]

pitchers_data.head()

pitch_type  release_speed  release_pos_x  release_pos_z pitcher p_throws  \
0         CU           84.9          -1.19           6.12  622250        R   
1         FF           96.6          -0.69           6.24  622250        R   
2         CU           84.5          -1.27           6.11  622250        R   
3         FF           95.4           -0.8           6.23  622250        R   
4         FF           95.6          -0.83           6.19  622250        R   

   release_spin_rate  effective_speed  inning  pitch_number mlbamid  \
0               2515             83.8       9             5     NaN   
1               2442             95.9       9             4     NaN   
2               2339             83.5       9             3     NaN   
3               2392             94.7       9             2     NaN   
4               2289             94.8       9             1     NaN   

  TJ Surgery Date  had_surgery  game_date  game_year  
0      No Surgery        False 2023-11-01       2023  
1      No Surgery        False 2023-11-01       2023  
2      No Surgery        False 2023-11-01       2023  
3      No Surgery        False 2023-11-01       2023  
4      No Surgery        False 2023-11-01       2023

## Now we have pitching characteristics data for all MLB Pitchers and an indication of whether or not they have had TJ surgery. 
## I want to create a model that predicts TJ surgery (UCL injury) based on this information.

In [842]:
import numpy as np

# Identify non-numeric columns
non_numeric_columns = pitchers_data.select_dtypes(exclude=[np.number]).columns
print("Non-numeric columns:", non_numeric_columns)

Non-numeric columns: Index(['pitch_type', 'pitcher', 'p_throws', 'mlbamid', 'TJ Surgery Date',
       'had_surgery', 'game_date'],
      dtype='object')


## Encoding categorical variables

In [888]:
#Encoding variables

# Mapping dictionary for pitch_type
pitch_type_mapping = {
    'FC': 0,
    'CU': 1,
    'FS': 2,
    'FF': 3,
    'SL': 4,
    'SI': 5,
    'CH': 6,
    'PO': 7,
    'ST': 8,
    'KC': 9,
    'FO': 10,
    'FA': 11,
    'SV': 12,
    'SC': 13,
    'CS': 14,
    'EP': 15
}

# Add numerical pitch_type column to the DataFrame
pitchers_data.loc[:, 'pitch_type_numeric'] = pitchers_data['pitch_type'].map(pitch_type_mapping)

# Mapping dictionary for p_throws
p_throws_mapping = {
    'R': 0,
    'L': 1
}

#Turn Bool had_surgery numeric
pitchers_data['had_surgery_numeric'] = pitchers_data['had_surgery'].astype(int)


pitchers_data


/var/folders/mj/b8_lt8gn6cvcflsf00z0pn6r0000gn/T/ipykernel_4367/339827784.py:30: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  pitchers_data['had_surgery_numeric'] = pitchers_data['had_surgery'].astype(int)


pitch_type  release_speed  release_pos_x  release_pos_z pitcher  \
0              CU           84.9          -1.19           6.12  622250   
1              FF           96.6          -0.69           6.24  622250   
2              CU           84.5          -1.27           6.11  622250   
3              FF           95.4           -0.8           6.23  622250   
4              FF           95.6          -0.83           6.19  622250   
...           ...            ...            ...            ...     ...   
771052       None           <NA>           <NA>           <NA>  641540   
771053       None           <NA>           <NA>           <NA>  641540   
771054       None           <NA>           <NA>           <NA>  641540   
771055       None           <NA>           <NA>           <NA>  641540   
771056       None           <NA>           <NA>           <NA>  641540   

       p_throws  release_spin_rate  effective_speed  inning  pitch_number  \
0             R               2515             83.8       9             5   
1             R               2442             95.9       9             4   
2             R               2339             83.5       9             3   
3             R               2392             94.7       9             2   
4             R               2289             94.8       9             1   
...         ...                ...              ...     ...           ...   
771052        R               <NA>             <NA>       1             1   
771053        R               <NA>             <NA>       1             3   
771054        R               <NA>             <NA>       1             2   
771055        R               <NA>             <NA>       1             1   
771056        R               <NA>             <NA>       1             1   

       mlbamid      TJ Surgery Date  had_surgery  game_date  game_year  \
0          NaN           No Surgery        False 2023-11-01       2023   
1          NaN           No Surgery        False 2023-11-01       2023   
2          NaN           No Surgery        False 2023-11-01       2023   
3          NaN           No Surgery        False 2023-11-01       2023   
4          NaN           No Surgery        False 2023-11-01       2023   
...        ...                  ...          ...        ...        ...   
771052  641540  2019-03-18 00:00:00         True 2023-03-15       2023   
771053  641540  2019-03-18 00:00:00         True 2023-03-15       2023   
771054  641540  2019-03-18 00:00:00         True 2023-03-15       2023   
771055  641540  2019-03-18 00:00:00         True 2023-03-15       2023   
771056  641540  2019-03-18 00:00:00         True 2023-03-15       2023   

        pitch_type_numeric  p_throws_numeric  had_surgery_numeric  
0                      1.0                 0                    0  
1                      3.0                 0                    0  
2                      1.0                 0                    0  
3                      3.0                 0                    0  
4                      3.0                 0                    0  
...                    ...               ...                  ...  
771052                 NaN                 0                    1  
771053                 NaN                 0                    1  
771054                 NaN                 0                    1  
771055                 NaN                 0                    1  
771056                 NaN                 0                    1  

[771057 rows x 18 columns]

In [906]:
# Group by some categorical variable (e.g., player name) and calculate totals for inning and pitch_number
totals = pitchers_data.groupby('pitcher')[['inning', 'pitch_number']].sum().reset_index()

# Rename the columns to indicate that they represent totals
totals = totals.rename(columns={'inning': 'total_innings', 'pitch_number': 'total_pitch_count'})

# Merge the totals DataFrame with the original DataFrame based on the categorical variable
pitchers_data = pd.merge(pitchers_data, totals, on='pitcher', how='left')

pitchers_data

pitch_type  release_speed  release_pos_x  release_pos_z pitcher  \
0              CU           84.9          -1.19           6.12  622250   
1              FF           96.6          -0.69           6.24  622250   
2              CU           84.5          -1.27           6.11  622250   
3              FF           95.4           -0.8           6.23  622250   
4              FF           95.6          -0.83           6.19  622250   
...           ...            ...            ...            ...     ...   
771052       None           <NA>           <NA>           <NA>  641540   
771053       None           <NA>           <NA>           <NA>  641540   
771054       None           <NA>           <NA>           <NA>  641540   
771055       None           <NA>           <NA>           <NA>  641540   
771056       None           <NA>           <NA>           <NA>  641540   

       p_throws  release_spin_rate  effective_speed  inning  pitch_number  \
0             R               2515             83.8       9             5   
1             R               2442             95.9       9             4   
2             R               2339             83.5       9             3   
3             R               2392             94.7       9             2   
4             R               2289             94.8       9             1   
...         ...                ...              ...     ...           ...   
771052        R               <NA>             <NA>       1             1   
771053        R               <NA>             <NA>       1             3   
771054        R               <NA>             <NA>       1             2   
771055        R               <NA>             <NA>       1             1   
771056        R               <NA>             <NA>       1             1   

        ... game_year pitch_type_numeric  p_throws_numeric  \
0       ...      2023                1.0                 0   
1       ...      2023                3.0                 0   
2       ...      2023                1.0                 0   
3       ...      2023                3.0                 0   
4       ...      2023                3.0                 0   
...     ...       ...                ...               ...   
771052  ...      2023                NaN                 0   
771053  ...      2023                NaN                 0   
771054  ...      2023                NaN                 0   
771055  ...      2023                NaN                 0   
771056  ...      2023                NaN                 0   

       had_surgery_numeric  total_innings_x  total_pitch_count_x  \
0                        0             7584                 2924   
1                        0             7584                 2924   
2                        0             7584                 2924   
3                        0             7584                 2924   
4                        0             7584                 2924   
...                    ...              ...                  ...   
771052                   1            11223                 8237   
771053                   1            11223                 8237   
771054                   1            11223                 8237   
771055                   1            11223                 8237   
771056                   1            11223                 8237   

        total_innings_y  total_pitch_count_y  total_innings  total_pitch_count  
0                  7584                 2924           7584               2924  
1                  7584                 2924           7584               2924  
2                  7584                 2924           7584               2924  
3                  7584                 2924           7584               2924  
4                  7584                 2924           7584               2924  
...                 ...                  ...            ...                ...  
771052            11223                 8237          11223     

In [922]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
from sklearn.impute import SimpleImputer



In [908]:
pitchers_data.columns

Index(['pitch_type', 'release_speed', 'release_pos_x', 'release_pos_z',
       'pitcher', 'p_throws', 'release_spin_rate', 'effective_speed', 'inning',
       'pitch_number', 'mlbamid', 'TJ Surgery Date', 'had_surgery',
       'game_date', 'game_year', 'pitch_type_numeric', 'p_throws_numeric',
       'had_surgery_numeric', 'total_innings_x', 'total_pitch_count_x',
       'total_innings_y', 'total_pitch_count_y', 'total_innings',
       'total_pitch_count'],
      dtype='object')

In [912]:
feature_columns = ['release_speed', 'release_pos_x', 'release_pos_z', 'release_spin_rate', 
                   'effective_speed', 'pitch_type_numeric', 'total_innings_x', 'total_pitch_count_x', 'p_throws_numeric']
                  

In [924]:
# Select features and target variable
X = pitchers_data[feature_columns]
y = pitchers_data['had_surgery_numeric']

# Impute missing values in the feature matrix
imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X)



In [926]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)


In [928]:
# Initialize and train the logistic regression model
model = LogisticRegression()
model.fit(X_train, y_train)


LogisticRegression()

In [930]:
# Make predictions on the testing data
y_pred = model.predict(X_test)



In [932]:
# Evaluate the model
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.6187067154307058
Classification Report:
               precision    recall  f1-score   support

           0       0.62      1.00      0.76     95403
           1       0.71      0.00      0.00     58809

    accuracy                           0.62    154212
   macro avg       0.67      0.50      0.38    154212
weighted avg       0.66      0.62      0.47    154212



## Breakdown of the classification report metrics:

Precision: Precision measures the accuracy of positive predictions. The precision for class "1" is relatively low (0.71), indicating that the model makes many false positive predictions.

Recall: Recall measures the ratio of positive instances that are correctly detected by the classifier. The recall for class "1" is also low (0.00), indicating that the model misses a significant number of positive instances.

F1-score: The F1-score is the harmonic mean of precision and recall. It provides a balance between precision and recall. 

F1-score for class "1" is also low (0.00), reflecting the poor performance of the model in detecting positive instances.
